Block producer refactor and chain bus integration

leventeliu opened the pull request:
In this new design, we use an immutable database to keep track of the chain state upon last irreversible block. Each optional brach keeps a runtime cache of its local state. Branches will grow, fork, or switch between one each other, driven by block producing and synchronization, and transactions will apply to the branch states.

Theoretically, block producers should maintain an eventual consistency chain given that the preset confirmation meets BFT requirement.

Chain bus model is added to chain structure for further use.

Commit:
Fix typo in log import

Commit:
Merge branch 'develop' of github.com:CovenantSQL/CovenantSQL into develop

Commit:
Prune useless chain pkg code

Commit:
Fix PersistentCaller Reconnect logic

Commit:
Add chainbus prototype

Commit:
Format imports

Commit:
Change topic to string type

Commit:
Make cql-fuse default log level to InfoLevel

Commit:
Fix chainbus test

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/chainBus

Commit:
Utils/Profiler log field name wrong.

Should log memprofile filename in memprofile section.

zeqing-guo committed:
Merge branch 'feature/chainBus' of github.com:CovenantSQL/CovenantSQL into feature/chainBus

zeqing-guo committed:
Merge branch 'feature/chainBus' of github.com:CovenantSQL/CovenantSQL into feature/chainBus

leventeliu committed:
Add block cache pruning

zeqing-guo committed:
Change AddTx to chainbus

zeqing-guo committed:
Move blockproducer/types to types and change the account fields

zeqing-guo committed:
Format code

zeqing-guo committed:
Merge branch 'feature/chainBus' of github.com:CovenantSQL/CovenantSQL into feature/chainBus

zeqing-guo committed:
Add miner-user matching mode

zeqing-guo committed:
Add update permission

leventeliu committed:
Use context instead of stop signal channel

leventeliu committed:
Use context version of CallNode

leventeliu committed:
Refactor imports

zeqing-guo committed:
Add test case

leventeliu committed:
Fix DATARACE in testing

zeqing-guo committed:
Support db encryption on SQLChainProfile

leventeliu committed:
Simplify goFunc of chain runtime

zeqing-guo committed:
Merge branch 'feature/chainBus' of github.com:CovenantSQL/CovenantSQL into feature/chainBus

leventeliu committed:
Use defer to decrement WaitGroup counter

zeqing-guo committed:
Format code

zeqing-guo committed:
Merge branch 'feature/chainBus' of github.com:CovenantSQL/CovenantSQL into feature/chainBus

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Add branch and runtime state

leventeliu committed:
Add method for both block producing and applying to switch branch

leventeliu committed:
Move state cache to metaState

leventeliu committed:
Remove boltdb, use new version block producing

leventeliu committed:
Remove main metaState in chain struct

leventeliu committed:
Remove chain testing

leventeliu committed:
Remove redundant codes in the new design

leventeliu committed:
Remove block index

leventeliu committed:
Add methods to reload chain state

leventeliu committed:
Refactor runtime constructor

leventeliu committed:
Add block node testing

leventeliu committed:
Add basic testing, fix some issues

leventeliu committed:
Fix hash stringer

leventeliu committed:
Add short string for clearer logging

leventeliu committed:
Use common config to run local node testing

leventeliu committed:
Fix missing base account issue in initialization

leventeliu committed:
Add testing cases and debug log

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Fix comment format

covenantdb commented on types/issuekeys.go:
```diff
@@ -0,0 +1,78 @@
+
+//go:generate hsp
+
+type MinerKey struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MinerKey should have comment or be unexported

covenantdb commented on types/issuekeys.go:
```diff
@@ -0,0 +1,78 @@
+	EncryptionKey string
+}
+
+type IssueKeysHeader struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type IssueKeysHeader should have comment or be unexported

covenantdb commented on xenomint/types/block.go:
```diff
@@ -54,7 +54,7 @@ func (h *SignedBlockHeader) Verify() error {
 }
 
-// Block defines a block including a signed block header and its query list.
+// BPBlock defines a block including a signed block header and its query list.
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported type Block should be of the form "Block ..." (with optional leading article)

leventeliu committed:
Fix issue at database reloading

leventeliu committed:
Fix an issue on branch forking

Check all branches for growing first. This will avoid to fork unnecessary
new branch.

leventeliu committed:
Refactor testing case

leventeliu committed:
Remove deprecated billing codes

leventeliu committed:
Add testing for chain APIs

leventeliu committed:
Minor fix for go linter

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Fix mixin type mismatching

leventeliu committed:
Rename rt to runtime

leventeliu committed:
Add a convey testing branch

This covers storage updating and reloading of some other state objects.

leventeliu committed:
Use block height to discover genesis

This avoids some problems in old testing data (such as using a
0x....0001 root hash).

leventeliu committed:
Avoid to create new branch on existed block

leventeliu committed:
Fetch head blocks parallelly

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Add RPC broadcast and context with timeout

leventeliu committed:
Add single run testing branch

leventeliu committed:
Fix transaction sorting: should be by account and nonce

leventeliu committed:
Set log level in testing mode to reduce log length

leventeliu committed:
Fix chain head synchronizing

leventeliu committed:
Remove root context

leventeliu committed:
Merge runtime struct to Chain

leventeliu committed:
Add transactions to forked branch

leventeliu committed:
Fix transaction type error

leventeliu committed:
Rename some functions to improve readability

auxten commented on blockproducer/storage.go:
```diff
@@ -0,0 +1,521 @@
+		return errPass(err)
+	}
+	return func(tx *sql.Tx) (err error) {
+		_, err = tx.Exec(`INSERT OR REPLACE INTO "provider" VALUES (?, ?)`,
```
Put column names in the query is better.

auxten commented on blockproducer/chain.go:
```diff
@@ -274,497 +292,696 @@ func (c *Chain) checkBlock(b *pt.Block) (err error) {
+	for _, b := range newIrres {
+		for _, tx := range b.block.Transactions {
+			if err := c.immutable.apply(tx); err != nil {
+				log.WithError(err).Fatal("Failed to apply block to immutable database")
```
Error or Fatal strings should not be capitalized

auxten commented on blockproducer/chain.go:
```diff
@@ -45,224 +49,238 @@ var (
+
+		peers:      cfg.Peers,
+		nodeID:     cfg.NodeID,
+		comfirms:   uint32(m),
```
m is a float64, uint32(m) may cause some vuln problems.

xq262144 replied:
Spelling mistake: `comfirms`  -> `confirms`

leventeliu committed:
Remove unused codes

leventeliu committed:
Format logging messages

find . -name '*.go' ! -path './vendor/*' ! -path './test/GNTE/*' -exec sed -i \
    's/\.\(Debug\|Info\|Warn\|Warning\|Error\|Fatal\|Panicf\?\)("\([[:upper:]]\)\([[:lower:]]\+[^[:alnum:]]\)/.\1("\L\2\3/' {} +

leventeliu committed:
Add column names in insert statement

leventeliu committed:
Fix typo

leventeliu committed:
Set testing log level

leventeliu committed:
Remove transaction pool codes

leventeliu committed:
Remove blank files

leventeliu committed:
Minor fix

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Minor fix

leventeliu committed:
Remove more unused codes

xq262144 commented on blockproducer/config.go:
```diff
@@ -30,21 +30,22 @@ const (
-	NodeID proto.NodeID
+	Peers            *proto.Peers
+	NodeID           proto.NodeID
+	ComfirmThreshold float64
```
Typo here.

leventeliu committed:
Fix typo

leventeliu committed:
Rename ambiguous variables

leventeliu committed:
Minor fix

leventeliu committed:
Rename more things for readability

leventeliu committed:
More renaming: use more specific function names

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/bpRefactor

leventeliu committed:
Merge branch 'develop' into feature/bpRefactor

auxten commented on types/init_service_type.go:
```diff
@@ -31,12 +31,13 @@ type InitService struct {
+	Memory              uint64                 // reserved memory in bytes
+	LoadAvgPerCPU       uint64                 // max loadAvg15 per CPU
+	EncryptionKey       string                 `hspack:"-"` // encryption key for database instance
+	EventualConsistency bool                   `hspack:"-"` // TODO, eventual consistency test, need later refactor
```
ConsistencyLevel float64 may be better

xq262144 replied:
Ok, good idea.

leventeliu replied:
This is a TODO from @xq262144 , which is currently implemented by calling `chain.Query` directly if the flag is set, i.e., the weakest eventual consistency. Need some more tinking in `kayak` later.

zeqing-guo committed:
Add test case for IssueKeys in metastate.go

zeqing-guo committed:
Merge branch 'feature/bpRefactor' of github.com:CovenantSQL/CovenantSQL into feature/bpRefactor

zeqing-guo committed:
Improve import format

Commit:
Use adjustable consistency level config instead of boolean eventual consistency config

leventeliu committed:
Add a FIXME and comment out channel closing to fix a DATARACE

Commit:
Add UseEventualConsistency flag for eventual consistency and use ConsistencyLevel for strong consistency confirmation setting

xq262144 reviewed (approved):
Approve
